### ISSUE TITLE
Add support for zathura as pdf viewer

### ISSUE THREAD

**Issue by lervag** [lervag/vim-latex#118]
Should be rather straightforward to implement full support based on the information provided by @Konfekt in #117:

```vimL
let g:latex_view_general_viewer = 'zathura'
let g:latex_view_general_options = shellescape('--synctex-forward "' . exepath(v:progpath) . ' --servername ' . v:servername . ' +{%lline} {%ffile}"')
nnoremap <expr><buffer><silent> gb  ':!zathura --synctex-forward '.line(".").':'.col('.').':% ' . shellescape(g:latex#data[b:latex.id].out()) . ' >/dev/null<CR>'
```

Also see [here](http://wiki.math.cmu.edu/iki/wiki/tips/20140310-zathura-fsearch.html).

**Comment by Konfekt** [comment]
Yep, there's (at least) one typo though:
```
let g:latex_view_general_options = shellescape('--synctex-forward "' . exepath(v:progpath) . ' --servername ' . v:servername . ' +{%line} {%file}"')
```
Also, `v:progname` instead of `v:progpath` works since Vim6, but forces Vim to be in $PATH. On Linux this is no issue though, on Windows quite so.

**Comment by nasenatmer** [comment]
Nice to see you want to implement `zathura` support in `vim-latex`! I'd suggest to do the debugging here as the other issue is more generally about the `synctex` functionality.

First of all: **gb mapping** → this is very neat and it would be cool to have it mapped to the viewer from the `view_method` setting. In my case, however, vim is blanked after pressing `gb` and I have to press `Enter` to get back to the buffer. If I add `<C-L` to the end of the mapping t's okay:
```vim
nnoremap <expr><buffer><silent> gb  ':!zathura --synctex-forward '.line(".").':'.col('.').':% ' . shellescape(g:latex#data[b:latex.id].out()) . ' >/dev/null<CR> <C-L>'
```

Then: setting the `general_options` prevents zathura from starting (with both of your variants). `\lv` just does nothing. No message in status line, no viewer opening, although there is no instance of zathura running. `\li` yields the following output for lervag's:
![lervag](https://cloud.githubusercontent.com/assets/1492278/6229433/c6a49d20-b6b7-11e4-9c68-33edca3d5de5.png)

and Konfekt's: 
![konfekt](https://cloud.githubusercontent.com/assets/1492278/6229437/d4d49a44-b6b7-11e4-812f-185eb31a1d1f.png)

I'm not sure whether this is of any help but I hope so at least.

**Comment by Konfekt** [comment]
Hopa, I mixed the commands up. Try
```
let g:latex_view_general_options = shellescape("-s -x '" . exepath(v:progpath) . " --servername " . v:servername . " +{%line} {%input}'")
```
Does that allow forward search, that is Vim -> Zathura?

**Comment by nasenatmer** [comment]
Forward search works flawlessly by setting `latex_view_general_viewer = 'zathura'` and using the `gb` mapping as suggested in the first post.

No, this line doesn't work. Nothing happens when pressing `\lv`. `\li` outputs:
```sh
command: view
    zathura '-s -x '\''/usr/bin/vim --servername VIM +{%line} {%input}'\''' '/home/jakob/doc/uni/teacher-theorie_und_praxis/essay/essay.pdf'
```

**Comment by lervag** [comment]
Thanks for the updates. My plan is to implement the functionality to be similar to `mupdf`, so that most of the desired settings and mappings are automatically set. It is very helpful to get the necessary options for the general viewer, since these will be the basis for the new viewer.

**Comment by Konfekt** [comment]
I am sorry, meant inverse search. But if it works like for Sumatrapdf it should be invoked by double clicking the pdf document. That said, the escaping by shellescape looks weird. Perhaps try also without it.

**Comment by Konfekt** [comment]
By the way, just realizing that these settings should also go into latexmkrc,  or better into vim-latex 's latexmkrc. If not, inverse search only works if vim starts the pdf viewer,  but not when latexmk.

**Comment by lervag** [comment]
@Konfekt I opened an issue for that: #119. I would be very happy if you could list the specific options we should set (though if you don't, I will still easily find these in the current issue).

**Comment by lervag** [comment]
The latest commit should fix some minor bugs. The implementation works well for me, but there are some annoying kinks:

1. If zathura is started by latexmk, then the inverse search must be set manually in ones `.latexmkrc` file
2. If zathura is started by latexmk, then `\lv` will open a new viewer before it starts forward searching

I have an idea to create a custom latexmkrc file before starting latexmk, which may be used to solve these issues, see #119. In any case, I think we need an improvement to the coupling between latexmk and vim-latex, and I think this is a large job and it might not be done in the nearest future. :\

**Comment by Konfekt** [comment]
Ok, done right its perhaps a big job, but these two points are exactly the concern at
https://github.com/lervag/vim-latex/issues/119#issuecomment-74826281 that could be resolved by
```vim 
let g:latex_latexmk_options =
    \ " -e ".shellescape("zathura -s -x '" . exepath(v:progpath) . " --servername " . v:servername . " +{%line} {%input}';")
```
However, most probably this won't work due to some inscrutable shell escaping issues, best indeed to create a configuration file containing
```
zathura -s -x '" . exepath(v:progpath) . " --servername " . v:servername . " +{%line} {%input}';
```
that is passed to latexmk.

**Comment by lervag** [comment]
@Konfekt I tried something similar to that, but as you've predicted I was not able to get it to work due to shell escaping issues.

**Comment by nasenatmer** [comment]
Wow, just trying out the new implementation and it seems to work better than the old method (setting zathura as `view_general`! pressing `\lv` (with `$pdf_previewer` and `$new_viewer_always=0` set in `.latexmkrc`) now still opens up a new instance of zathura (although one has already been opened by `latexmk`) **but** the old one is marked by synctex. This wasn't the case before. Cool!

**Comment by lervag** [comment]
In the final version pushed just now, `\lv` should not open a new instance of zathura if one was already opened with `latexmk`. However, the opposite is not true, and I don't see any solution for that. That is: If I open the pdf with `\lv`, then start compilation with `\ll`, a second viewer is opened by `latexmk`, regardless of the value of the `$new_viewer_always` option.

**Comment by nasenatmer** [comment]
**WOOOOW**, nice one! Indeed, the one zathura instance opened by `\ll` is the only one opening also after pressing `\lv` several times! Really cool! Also, backward search works out of the box now! Amazing!

One question: zathura isn't focused when pressing `\lv`. But vim stays focused. Same is true the other way for backward search: zathura stays focused while the cursor position is changing in vim. Is this intended?

**Comment by nasenatmer** [comment]
Working on a tex file since a few hours I just noticed that `vim-latex` is telling me in vim statusline whether compilation was successful, which is a really great substitute for xdotools auto-updating pdf-window title! Great!

**Comment by lervag** [comment]
When I found out that the window title in mupdf could not be opened if I wanted to enable forward search, I realized that we needed a substitute. I found that highlighted statusline messages worked well. I'm glad you agree! :)

And yes: I added an `xdotool` command to return focus to vim after `\lv`, similar to what I did for mupdf. I have not really decided if I think this is more natural or not, and I could change it if desired (or add an option for it).

**Comment by lervag** [comment]
When I CTRL-click in zathura for inverse search, focus is also returned to vim as expected.

**Comment by nasenatmer** [comment]
Strange. I see the same behaviour on my work machine as on my home computer (though I expect this to be the case as I sync my dotfiles via github).

But then again this is probably under control of the window manager, because `vim-latex` only sets up the relation between zathura and vim whereas the decision to focus vim or keep the focus in the zathura window after `ctrl-click` is eihter `zathuras` or the window manager's. (Openbox on both my machines…).

**Comment by lervag** [comment]
Yes, that is strange. I agree that `ctrl-click` in `zathura` should be handled either by `zathura` or by your window manager. I use xfce4.

Regarding my decision to keep focus in vim after starting mupdf or zathura: do you agree with this choice, or is it more natural to change focus? @Konfekt @nasenatmer

**Comment by nasenatmer** [comment]
I think it depends on the case:

1. pdf viewer is opened by latexmk automatically: focus should stay with vim as one might not expect the viewer to pup up (but actually, this is open to discussion, I've only been using this auto-opening function since yesterday)

2. pdf viewer is opened by `\lv` or forward-search is executed: Focus should switch to zathura as I'd usually would like to check how it looks there.

A config option might be nice although we seem to be accumulating them recently. Is there a way to config this in the user's `~/.latexmkrc` in order to override `vim-latex` default behaviour? It seems `vim-latex` has this function via xdotools…

**Comment by lervag** [comment]
The current version overrides the `$pdf_previewer` setting of `.latexmkrc`. I agree that this should be optional, so I'll create a new option for that. But perhaps you rather meant to ask if it is possible to set the focus option thing in `.latexmkrc`? In that case: I don't think so. `vim-latex` uses xdotools to change focus.

Further, I agree that for pdf viewer opened by `latexmk` we should return the focus, while `\lv` should change focus. I'll make the change. I'll prefer not to add an option for this, and rather to find the sensible default. Unless you all disagree?

**Comment by nasenatmer** [comment]
To me, these a second-order concerns, so I'm happy either way. As long as the behaviour is consistent and predictible (i.e. stable), I'm fine.

Yes: I meant to say that instead of creating more and more config options it would be an alternative to provide the necessary documentation for users who want to "voluntarily" change things from the defaults if they feel like it. But if this is in no relation to the rising complexity, I think it's not worth it…

TL;DR: sensible defaults :+1:

**Comment by Konfekt** [comment]
Nice commit @lervag, thanks for all the hard work.

**Comment by lervag** [comment]
No problem. Thanks for all the input and the good discussions! I've pushed some changes that should set more sensible defaults. Please open new issues if you have feedback to the new code, be it bugs or comments to the default settings or similar.

**Comment by msaidf** [comment]
the backward search is still not working out of the box for me. I tried Ctrl-click on zathura, but nothing happened.  Do I need to put anything manually on zathurarc or latexmkrc to make it work? I have put let g:vimtex_view_method='zathura' on my vimrc and get the forward search working

**Comment by lervag** [comment]
That is strange. Could you open a new issue for this. Please describe:
1. Expected behaviour (like you have done here).
2. Which vim version you have installed.
3. Which zathura version you have installed.
4. Which settings you use for vimtex.

**Comment by msaidf** [comment]
Sorry, I should have mentioned that I was using Neovim. When I tried on Vim, it is working fine. But if you don't mind, I will open new issue for neovim compatibility. Thanks.

**Comment by scottbaron** [comment]
➜  ~  brew install gtk+3
==> Downloading https://download.gnome.org/sources/gtk+/3.16/gtk+-3.16.4.tar.xz
Already downloaded: /Library/Caches/Homebrew/gtk+3-3.16.4.tar.xz
==> ./configure --prefix=/usr/local/Cellar/gtk+3/3.16.4 --disable-glibtest --ena
==> make install
gtksettings.c:910: Warning: Gtk: unexpected annotation: element-type
Command '['/private/tmp/gtk+320150622-97160-asjxi8/gtk+-3.16.4/gtk/tmp-introspectKvgQzd/Gtk-3.0', '--introspect-dump=/private/tmp/gtk+320150622-97160-asjxi8/gtk+-3.16.4/gtk/tmp-introspectKvgQzd/functions.txt,/private/tmp/gtk+320150622-97160-asjxi8/gtk+-3.16.4/gtk/tmp-introspectKvgQzd/dump.xml']' returned non-zero exit status -11
make[2]: *** [Gtk-3.0.gir] Error 1
make[1]: *** [install] Error 2
make: *** [install-recursive] Error 1

READ THIS: https://git.io/brew-troubleshooting

These open issues may also help:
gtk+3: adwaita-icon-theme added as resource (https://github.com/Homebrew/homebrew/pull/40888)
Gtk+3 fails to compile on 10.6.8 (https://github.com/Homebrew/homebrew/issues/40438)
 ✘ ➜  ~ 

**Comment by scottbaron** [comment]
Trying to install Zathura on MACOS 10.6.8 with brew getting an error trying to install GTK+3 on the build getting the following error:➜  ~  brew install gtk+3
==> Downloading https://download.gnome.org/sources/gtk+/3.16/gtk+-3.16.4.tar.xz
Already downloaded: /Library/Caches/Homebrew/gtk+3-3.16.4.tar.xz
==> ./configure --prefix=/usr/local/Cellar/gtk+3/3.16.4 --disable-glibtest --ena
==> make install
gtksettings.c:910: Warning: Gtk: unexpected annotation: element-type
Command '['/private/tmp/gtk+320150622-97160-asjxi8/gtk+-3.16.4/gtk/tmp-introspectKvgQzd/Gtk-3.0', '--introspect-dump=/private/tmp/gtk+320150622-97160-asjxi8/gtk+-3.16.4/gtk/tmp-introspectKvgQzd/functions.txt,/private/tmp/gtk+320150622-97160-asjxi8/gtk+-3.16.4/gtk/tmp-introspectKvgQzd/dump.xml']' returned non-zero exit status -11
make[2]: *** [Gtk-3.0.gir] Error 1
make[1]: *** [install] Error 2
make: *** [install-recursive] Error 1

READ THIS: https://git.io/brew-troubleshooting

These open issues may also help:
gtk+3: adwaita-icon-theme added as resource (https://github.com/Homebrew/homebrew/pull/40888)
Gtk+3 fails to compile on 10.6.8 (https://github.com/Homebrew/homebrew/issues/40438)
 ✘ ➜  ~ 

**Comment by nasenatmer** [comment]
@scottbaron That seems to be a problem between  zathura/gtk3 und brew which we can hardly solve here.

**Comment by lervag** [comment]
@scottbaron As @nasenatmer comments, this is not an issue with vimtex. I don't personally use OSX and Homebrew, and I am therefore not able to help you with this. I'm sorry for the inconvenience.

**Comment by ktran9891** [comment]
Sorry to raise the dead here, but would it be possible to get a step-by-step summary of what a new user needs to do to use Zathura?

I am confused as to which of the multiple proposed solutions to use in the multiple issue threads I've found, and which of these solutions work with which versions of vimtex.

And for my personal (and more selfish) context:  I am using vim (i.e., not gvim) on OS X. So does this mean that I need to open the file with a `vim --servername foo file.tex`? Also, is it safe to assume that I can put any text in `foo`, or do I need to set up some third party server somewhere?

**Comment by andymass** [comment]
@ktran9891   It *should* be as simple as `let g:vimtex_view_method = 'zathura'` (see example vimrc below).  Ensure zathura is in your path.

Yes, you need to use `vim --servername foo file.tex` because vim does not default to a servername in terminal.  foo can be just about anything you like (but it's best to stick to letters only and no spaces).  The easiest way is to make an alias in your shell.  If you create multiple vim instances with the same --servername vim will add a number so as to make it unique.

```
set nocompatible

" Load Vimtex
let &rtp  = '~/.vim/bundle/vimtex,' . &rtp
let &rtp .= ',~/.vim/bundle/vimtex/after'

" Load other plugins, if necessary
" let &rtp = '~/path/to/other/plugin,' . &rtp

filetype plugin indent on
syntax enable

" Vimtex options go here
let g:vimtex_view_method = 'zathura'
```

**Comment by trimitri** [comment]
I can confirm that just setting ``let g:vimtex_view_method = 'zathura'`` works out of the box in Arch Linux; forward-search, backward-search, auto-update.  Vimtex notifies you, though, that you need to install _xdotool_, which takes two seconds.

Thanks for the great work on this plugin!
